### PR TITLE
Improve diagnostics when a static lifetime is expected

### PR DESCRIPTION
**Pull request by rukai**
Makes progress towards https://github.com/rust-lang/rust/issues/90600

The diagnostics here were previously entirely removed due to giving a misleading suggestion but if we instead provide an informative label in that same location it should better help the user understand the situation.

I included the example from the issue as it demonstrates an area where the diagnostics are still lacking.
Happy to remove that if its just adding noise atm.

### CODE REVIEW

**Comment by michaelwoerister** [comment]
r? rust-lang/diagnostics

**Comment by estebank** [comment]
As you note, the next step is to point at the source of the requirement.

r? @estebank

@bors r+

**Comment by matthiaskrgr** [comment]
Looks like there are some test failures. Please check
https://github.com/rust-lang-ci/rust/runs/4198301917?check_suite_focus=true#step:25:32013
@bors r-

**Comment by rukai** [comment]
Ah, so I need to specify `--compare-mode nll` to run those failed tests, thats a little hidden.
I guess I'll take this opportunity to replace this PR with the fix I mentioned in https://github.com/rust-lang/rust/issues/90600#issuecomment-967797354 as I think this PR as it is currently is a bit of a misstep.

**Comment by rukai** [comment]
Ready for review, the PR is completely rewritten so you should rereview from scratch.
Now we are falling back to other errors when suggesting adding an explicit lifetime would be inappropriate.

I am currently running the following to ensure the tests are passing:
```bash
RUST_BACKTRACE=0 ./x.py test --stage 1 --force-rerun --compare-mode nll
RUST_BACKTRACE=0 ./x.py test --stage 1 --force-rerun
```

# Observed problems in the new diagnostics.

## 1
nll is missing the "captured here" notes that migrate gives, e.g.
```
generator-region-requirements.rs
+LL |         x
+   |         ^ ...is captured here...
```
## 2
The nll diagnostics always name the lifetimes (e.g. `1) even when there is only one lifetime.
Assigning a name to something is an extra mental burden, so it would be a lot easier to read if we didnt do that when there is only one lifetime.

## 3
The nll diagnostics say:
```
returning this value requires that `'1` must outlive `'static`
```

Considering you cant outlive `static, in this case it might make more sense to say:

```
returning this value requires that `'1` must be `'static`
```

But it seems none of the diagnostics are regressing in value so any potentital fixes for these can go in follow up PRs.

If the listed problems make sense to you then I'll file a seperate issue for each of them.

# Strange UI test
regions-static-bound has stderr files named like `regions-static-bound.migrate.nll.stderr` I have no idea what `.migrate.nll.stderr` could mean so I assumed it was broken and fiddled with the config to make it generate normal stderr files.
Let me know if I just broke something I didnt understand.

File: src/test/ui/regions/regions-static-bound.stderr [hunk 1]
```diff
@@ -0,0 +1,46 @@
+error[E0312]: lifetime of reference outlives lifetime of borrowed content...
+  --> $DIR/regions-static-bound.rs:6:5
+   |
+LL |     t
+   |     ^
+   |
+   = note: ...the reference is valid for the static lifetime...
```
Comment (estebank) [inline src/test/ui/regions/regions-static-bound.stderr:7]: It seems like we're not looking at the bounds in the `where` clause to look for the source of `'static` requirement. Not necessary on this PR.

**Review by estebank — APPROVED** [review]

**Pull request merged**